### PR TITLE
Allow executing CLI worker in parallel

### LINKED ISSUE #3165: Parallelize CPU bound workers
There are some workers that are heavily CPU bound, such as image processing, that could benefit from some degree of parallelization. Especially in the server world most CPU SKUs are skewed towards core count rather than frequency.

### PR DESCRIPTION
**Pull request by TimWolla**
worker

<details><summary>Old Description</summary>

This is a very rough initial proof of concept for executing the WorkerCLICommand in parallel. What's missing is a proper synchronization with the parent process for a progress indicator.

It currently works like this: If the `worker` CLI command is executed with a `--threads=X` parameter the command will re-execute the `worker` command within X child processes with a different `threadId` each. To do so a `sessionID` parameter is added (not automatically used yet) that will avoid the need for the child processes to actually perform the authentication at the start of the CLI session.

The worker command is adjusted, so that if a thread ID and a number of threads is given each thread will only execute the loops with the number `(threads * n) + threadId + 1`. The initial iteration is always performed before starting the threads, because it usually is used to perform clean up actions that need to happen before anything is actually rebuilt.

I'm putting this PR up for early review of the concept. It will be cleaned up if you don't find any major issues with it.
</details>

-------

Resolves #3165

### CODE REVIEW

File: wcfsetup/install/files/lib/system/cli/command/WorkerCLICommand.class.php [hunk 1]
```diff
@@ -109,12 +124,64 @@ public function execute(array $parameters) {
...
+			}
+			while (!empty($processes)) {
+				$read = array_map(function ($process) {
+					return $process['pipes'][3];
+				}, $processes);
+				$write = null;
+				$except = null;
+				$ready = stream_select($read, $write, $except, null, null);
+				foreach ($read as $fd) {
+					var_dump(fgets($fd));
```
Comment (joshuaruesweg) [inline wcfsetup/install/files/lib/system/cli/command/WorkerCLICommand.class.php:161]: Debug Code

File: wcfsetup/install/files/lib/system/cli/command/WorkerCLICommand.class.php [hunk 2]
```diff
@@ -109,12 +124,64 @@ public function execute(array $parameters) {
...
+				}
+				foreach ($processes as $key => $process) {
+					$status = proc_get_status($process['process']);
+					if ($status === false) {
+						throw new Exception('Should not happen');
+					}
+					if (!$status['running']) {
+						while (!feof($process['pipes'][3])) var_dump(fgets($process['pipes'][3]));
```
Comment (joshuaruesweg) [inline wcfsetup/install/files/lib/system/cli/command/WorkerCLICommand.class.php:169]: Debug Code

File: wcfsetup/install/files/lib/system/cli/command/WorkerCLICommand.class.php [hunk 3]
```diff
@@ -109,12 +124,64 @@ public function execute(array $parameters) {
...
+				$write = null;
+				$except = null;
+				$ready = stream_select($read, $write, $except, null, null);
+				foreach ($read as $fd) {
+					var_dump(fgets($fd));
+				}
+				foreach ($processes as $key => $process) {
+					$status = proc_get_status($process['process']);
+					if ($status === false) {
+						throw new Exception('Should not happen');
```
Comment (joshuaruesweg) [inline wcfsetup/install/files/lib/system/cli/command/WorkerCLICommand.class.php:166]: The error message should be more descriptive. Maybe also `\LogicException`?

File: wcfsetup/install/files/lib/system/CLIWCF.class.php [hunk 4]
```diff
@@ -126,7 +132,8 @@ protected function initArgv() {
 			'version' => WCF::getLanguage()->get('wcf.cli.help.version'),
 			'disableUpdateCheck' => WCF::getLanguage()->get('wcf.cli.help.disableUpdateCheck'),
 			'exitOnFail' => WCF::getLanguage()->get('wcf.cli.help.exitOnFail'),
-			'packageID=i' => WCF::getLanguage()->get('wcf.cli.help.packageID')
+			'packageID=i' => WCF::getLanguage()->get('wcf.cli.help.packageID'),
+			'sessionID=s' => 'sessionid',
```
Comment (TimWolla) [inline wcfsetup/install/files/lib/system/CLIWCF.class.php:136]: Use language item

File: wcfsetup/install/files/lib/system/cli/command/WorkerCLICommand.class.php [hunk 5]
```diff
@@ -54,6 +57,19 @@ public function execute(array $parameters) {
...
 		}
 		
+		$threads = 1;
+		$threadId = null;
+		if ($this->argv->threads !== null) {
+			$threads = $this->argv->threads;
+			if ($this->argv->threadId !== null) {
+				$threadId = $this->argv->threadId;
+				if ($this->argv->threadId >= $this->argv->threads) {
+					CLIWCF::getReader()->println("Invalid threadId");
```
Comment (TimWolla) [inline wcfsetup/install/files/lib/system/cli/command/WorkerCLICommand.class.php:67]: Use language item

File: wcfsetup/install/files/lib/system/cli/command/WorkerCLICommand.class.php [hunk 6]
```diff
@@ -109,12 +125,26 @@ public function execute(array $parameters) {
 		$worker->validate();
 		$worker->getProgress(); // make sure objects are counted
 		
-		// initialize progressbar
+		// If multithreading is requested, but no threadId is given then this
+		// is the controller process invoked by the Administrator.
+		if ($threads > 1 && $threadId === null) {
+			if (PHP_EOL === "\r\n") {
+				CLIWCF::getReader()->println('The --threads option is not available on Windows.');
```
Comment (TimWolla) [inline wcfsetup/install/files/lib/system/cli/command/WorkerCLICommand.class.php:132]: Use language item

File: wcfsetup/install/files/lib/system/cli/command/WorkerCLICommand.class.php [hunk 7]
```diff
@@ -125,12 +155,174 @@ public function execute(array $parameters) {
...
+		
+		// Invoke the worker processes with the same command line ...
+		$arguments = $_SERVER['argv'];
+		// ... with the quiet argument ...
+		$arguments[] = '-qqqqq';
+		// ... reusing the current session.
+		$arguments[] = '--sessionID='.CLIWCF::getSession()->sessionID;
+		$commandLine = PHP_BINARY.' '.implode(' ', array_map('escapeshellarg', $arguments));
+		
+		CLIWCF::getReader()->println('Using "'.$commandLine.'" as the worker command line.');
```
Comment (TimWolla) [inline wcfsetup/install/files/lib/system/cli/command/WorkerCLICommand.class.php:201]: Remove this line / use language item

**Comment by TimWolla** [comment]
This PR is ready now. I realized that giving the sessionID as a command line argument is unsafe, because it can be seen within the process list. So it now is passed as an environment variable.

@joshuaruesweg Please test once again on macOS.

**Pull request merged**